hathor/mining/block_template.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2019 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Module for abstractions around generating mining templates.
+"""
+
+import random
+import time
+from typing import Dict, List, NamedTuple, Optional, Set, Tuple, Type, Union
+
+from hathor.transaction import BaseTransaction, Block, MergeMinedBlock
+from hathor.transaction.storage import TransactionStorage
+
+
+class BlockTemplate(NamedTuple):
+    versions: Set[int]
+    reward: int  # reward unit value, 64.00 HTR is 6400
+    weight: float  # calculated from the DAA
+    timestamp_min: int  # min valid timestamp
+    timestamp_max: int  # max valid timestamp
+    parents: List[bytes]  # required parents, will always have a block and at most 2 txs
+    parents_any: List[bytes]  # list of extra parents to choose from when there are more options
+    height: int  # metadata
+    score: float  # metadata
+
+    def generate_minimaly_valid_block(self) -> BaseTransaction:
+        """ Generates a block, without any extra information that is valid for this template. No random choices."""
+        from hathor.transaction import TxVersion, TxOutput
+        return TxVersion(min(self.versions)).get_cls()(
```

Comment:
This is partially inspired by Bitcoin's RPC and from feedback from EuroLine. I realized that returning a nearly complete "binary" (hex-encoded) header, some implementations get much easier. Bitcoin's RPC does this for transactions (on the list of transactions that make the merkle root) but not for the block header. I think we can do this for the block template if it helps clients choose this API instead of `/v1a/mining` or `/v1a/get_block_template`.

Reply:
Got it